allennlp/nn/decoding/maximum_marginal_likelihood.py

```diff
@@ -0,0 +1,103 @@
+from collections import defaultdict
+from typing import Dict, Tuple, Set
+
+import torch
+
+from allennlp.common import Params
+from allennlp.nn import util
+from allennlp.nn.decoding.decoder_step import DecoderStep
+from allennlp.nn.decoding.decoder_state import DecoderState
+from allennlp.nn.decoding.decoder_trainer import DecoderTrainer
+
+
+@DecoderTrainer.register('max_marginal_likelihood')
+class MaximumMarginalLikelihood(DecoderTrainer):
+    """
+    This class trains a decoder by maximizing the marginal likelihood of the targets.  That is,
+    during training, we are given a `set` of acceptable or possible target sequences, and we
+    optimize the `sum` of the probability the model assigns to each item in the set.  This allows
+    the model to distribute its probability mass over the set however it chooses, without forcing
+    `all` of the given target sequences to have high probability.  This is helpful, for example, if
+    you have good reason to expect that the correct target sequence is in the set, but aren't sure
+    `which` of the sequences is actually correct.
+
+    This implementation of maximum marginal likelihood requires the model you use to be `locally
+    normalized`; that is, at each decoding timestep, we assume that the model creates a normalized
+    probability distribution over actions.  This assumption is necessary, because we do no explicit
+    normalization in our loss function, we just sum the probabilities assigned to all correct
+    target sequences, relying on the local normalization at each time step to push probability mass
+    from bad actions to good ones.
+    """
+    def decode(self,
+               initial_state: DecoderState,
+               decode_step: DecoderStep,
+               targets: torch.Tensor = None,
+               target_mask: torch.Tensor = None) -> Dict[str, torch.Tensor]:
+        if targets.dim() != 2:
+            raise NotImplementedError("This implementation cannot yet handle batched inputs")
+        allowed_transitions = self._create_allowed_transitions(targets, target_mask)
```

Comment:
Why is constrained decoding an essential part of `MaximumMarginalLikelihood`? I think it makes more sense for a `DecoderState` to define `allowed_transitions`.

Reply:
Sorry, ignore the previous comment. I misunderstood what you're doing here. I now know that this is not related to constrained decoding.